### DISCUSSION TITLE
Can I add a custom factory method onto DateTime?

### DISCUSSION THREAD

**Discussion by yeoffrey** [moment/luxon#1600] [Q&A] [answered]
Suppose I had an DateTime class with custom functions on it like in this example [here](https://github.com/moment/luxon/discussions/1017). Now, I want to add a custom factory method that can take another class and create a DateTime instance based on certain characteristics. In our example, we are writing custom methods for converting Google's `BigQueryTimestamp` from and to a `DateTime` object. 

```typescript
import {BigQueryTimestamp} from '@google-cloud/bigquery';
import {DateTime} from 'luxon';

declare module 'luxon/src/datetime' {
    export interface DateTimeConstructor {
        fromBigQueryTimestamp(bqt: BigQueryTimestamp): DateTime;
    }

    export interface DateTime {
        fromBigQueryTimestamp(bqt: BigQueryTimestamp): DateTime;

        toBigQueryTimestamp(): BigQueryTimestamp;
    }
}

DateTime.prototype.toBigQueryTimestamp = function (
    this: DateTime,
): BigQueryTimestamp {
    return new BigQueryTimestamp(this.toJSDate());
};

// @ts-expect-error TS2339: Property 'fromBigQueryTimestamp' does not exist on type 'typeof DateTime'.
DateTime.fromBigQueryTimestamp = function (bqt: BigQueryTimestamp): DateTime {
    return DateTime.fromJSDate(new Date(bqt.value));
};

export {DateTime};
```

This works if we ignore the error on `DateTime.fromBigQueryTimestamp ...` 
Error: ```Property 'fromBigQueryTimestamp' does not exist on type 'typeof DateTime'.```

So we can do this:

```typescript
import {DateTime} from "personal/luxon";
import {BigQueryTimestamp} from "@google-cloud/bigquery";

const date = (DateTime as unknown as DateTime).fromBigQueryTimestamp(new BigQueryTimestamp(new Date()));
console.log(date);
```

So obviously `(DateTime as unknown as DateTime)` isn't great and thats what I'm trying to get around, but I can't figure out a way around this. The constructor for `DateTime` is private so I can't make modifications without this hacky method. Would be nice to be able to extend it somehow so if anyone can point me in the right direction I would appreciate it!

**Comment by diesieben07** [comment, marked as answer]
Personally, I would stay far away from mucking around with "other people's prototypes". It has caused nothing but grief in the past (one example: https://developer.chrome.com/blog/smooshgate).
Just write a plain old function:
```ts
function dateTimeFromBigQueryTimestamp(bqt: BigQueryTimestamp): DateTime {
   // whatever
}
```

This also makes it clear that this is your own function and not Luxon's, which has the advantage of making it clearer to understand for other people reading your code. They won't be expecting documentation for this function in Luxon's documentation, but yours.

**Reply by yeoffrey** [reply]
yeah that probably seems easiest. Thanks for the feedback!
